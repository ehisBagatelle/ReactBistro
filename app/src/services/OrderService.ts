import { taxes } from "../../test/fixture";
import { formatDecimal } from "../helper";

export type LineItem = {
  name: string;
  category: string;
  price: number;
};

export type Tax = {
  name: string;
  rate: number;
  appliesToCategories: string[];
};

export type Discount = {
  id: number
  type: "dollar" | "percentage";
  amount: number;
};

export type Bill = {
  totalTaxes: string;
  totalDiscounts: string;
  subTotal: string;
  total: string;
};

export interface IOrderService {
  calculateBill(lineItems: LineItem[], appliedDiscounts: Discount[]): Bill;
}

export class OrderService implements IOrderService {
  taxes: Tax[] = taxes;

  public calculateBill(
    lineItems: LineItem[],
    appliedDiscounts: Discount[]
  ): Bill {
    let subTotal = 0;

    // Calculate pre-tax total
    lineItems.forEach((item) => {
      subTotal += item.price;
    });

    // Calculate total taxes
    const totalTaxes = this.calculateTotalTaxes(lineItems);
    
    // Calculate post-tax total
    const postTaxTotal = subTotal + totalTaxes;

    // Calculate total discounts
    const totalDiscounts = this.calculateTotalDiscounts(
      appliedDiscounts,
      postTaxTotal
    );

    const total = postTaxTotal - totalDiscounts

    return {
      totalTaxes: formatDecimal(totalTaxes),
      totalDiscounts: formatDecimal(totalDiscounts),
      subTotal: formatDecimal(subTotal),
      total: formatDecimal(total)
    };
  }

  private calculateTotalTaxes(lineItems: LineItem[]): number {
    let totalTaxes = 0;

    lineItems.forEach((item) => {
      const applicableTaxes = this.taxes.filter((tax) =>
        tax.appliesToCategories.includes(item.category)
      );

      applicableTaxes.forEach((tax) => {
        totalTaxes += item.price * tax.rate;
      });
    });

    return totalTaxes;
  }

  private calculateTotalDiscounts(
    appliedDiscounts: Discount[],
    postTaxTotal: number
  ): number {
    // The order of applied discounts matter and can change the final output totals
    // We recommend going dollar amount discounts first
    let postDiscountTotal = postTaxTotal;
  
    // Separate dollar amount discounts and percentage discounts
    const dollarAmountDiscounts = appliedDiscounts.filter(
      (discount) => discount.type === "dollar"
    );
  
    const percentageDiscounts = appliedDiscounts.filter(
      (discount) => discount.type === "percentage"
    );
  
    // Apply dollar amount discounts first
    dollarAmountDiscounts.forEach((discount) => {
        postDiscountTotal -= discount.amount;
    });
  
    // Apply percentage discounts after dollar amount discounts
    percentageDiscounts.forEach((discount) => {
        postDiscountTotal -= (discount.amount / 100) * postDiscountTotal;
    });
  
    return postTaxTotal - postDiscountTotal;
  }
}
