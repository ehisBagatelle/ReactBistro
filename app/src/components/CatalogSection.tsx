import React from "react";
import { SectionList } from "react-native";
import CatalogItem from "./CatalogItem";
import { Title } from "../components/Title";
import { LineItem } from "../services/OrderService";

interface Props {
  sections: {
    title: string;
    data: LineItem[];
  }[];
  onCatalogItemPress: (item: LineItem) => void;
}

const CatalogSection: React.FC<Props> = ({ sections, onCatalogItemPress }) => {
  return (
    <SectionList
      sections={sections}
      keyExtractor={(item, index) => item.name + index}
      renderItem={({ item }) => (
        <CatalogItem item={item} onPress={() => onCatalogItemPress(item)} />
      )}
      renderSectionHeader={({ section: { title } }) => <Title name={title} />}
    />
  );
};

export default CatalogSection;
