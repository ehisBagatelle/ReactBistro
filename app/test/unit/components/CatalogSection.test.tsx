import React from 'react';
import renderer from 'react-test-renderer';
import CatalogSection from '../../../src/components/CatalogSection';
import { LineItem } from '../../../src/services/OrderService';

const mockOnCatalogItemPress = jest.fn();

const sampleItem: LineItem = {
  key: "0002",
  name: 'Test Item',
  category: 'Test Category',
  price: 9.99,
};

const sampleSections = [
  {
    title: 'Test Section 1',
    data: [sampleItem, sampleItem, sampleItem],
  },
  {
    title: 'Test Section 2',
    data: [sampleItem, sampleItem],
  },
];

describe('CatalogSection', () => {
  it('renders correctly with sections and items', () => {
    const tree = renderer.create(
      <CatalogSection sections={sampleSections} onCatalogItemPress={mockOnCatalogItemPress} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
