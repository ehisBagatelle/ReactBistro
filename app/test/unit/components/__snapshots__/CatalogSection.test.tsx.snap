// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogSection renders correctly with sections and items 1`] = `
<RCTScrollView
  data={
    [
      {
        "data": [
          {
            "category": "Test Category",
            "key": "0002",
            "name": "Test Item",
            "price": 9.99,
          },
          {
            "category": "Test Category",
            "key": "0002",
            "name": "Test Item",
            "price": 9.99,
          },
          {
            "category": "Test Category",
            "key": "0002",
            "name": "Test Item",
            "price": 9.99,
          },
        ],
        "title": "Test Section 1",
      },
      {
        "data": [
          {
            "category": "Test Category",
            "key": "0002",
            "name": "Test Item",
            "price": 9.99,
          },
          {
            "category": "Test Category",
            "key": "0002",
            "name": "Test Item",
            "price": 9.99,
          },
        ],
        "title": "Test Section 2",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    [
      0,
      5,
    ]
  }
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#F5F5F5",
            "display": "flex",
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <Text>
          Test Section 1
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          Test Item
        </Text>
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          $
          9.99
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          Test Item
        </Text>
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          $
          9.99
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          Test Item
        </Text>
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          $
          9.99
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#F5F5F5",
            "display": "flex",
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <Text>
          Test Section 2
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          Test Item
        </Text>
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          $
          9.99
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "opacity": 1,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          Test Item
        </Text>
        <Text
          style={
            {
              "fontSize": 18,
              "fontStyle": "normal",
              "fontWeight": "400",
            }
          }
        >
          $
          9.99
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;
